### DISCUSSION TITLE
Is airflow idempotent?

### DISCUSSION THREAD

**Discussion by aamster** [apache/airflow#28905] [Q&A]
In Luigi for example, if data already exists, then the task will not overwrite it. Therefore it is by design idempotent. 

This is important for scientific purposes where we don't want to modify data. 

It seems airflow does not behave similarly and will gladly overwrite data if you rerun a workflow. Is this the case or am I missing something?

**Comment by potiuk** [comment]
This is an interesting quesiton.

And Yes. You miss something - or rather your assumptions are different. You seem to understand idempotency as immutability which is just one way of how "idempotency' can be understood - and a very simplicitc one, not reflecting many real-world use cases.

Airflow operators are supposed to be idempotent and idempotency is understood more on a desired output - i.e. when you re-run it the result produced represent the expected output that potentially replaces the one that was generated before. So result is overwritten, possibly with updated/fixed result that better reflect current state of the world.

While "idempotent" is a nice word, it does not have a nice, universal scientific definition that can be encoded. And Airflow approaches "ideal" approach is to make it pragmatic and a little vague (but this is how real world behaves).

An example of that would be when you process list of calls made by a telecom and get the aggregate of phone models used. You typically map the IMEI of the phones to the models. But the mapping might be incomplete and might be updated after the data has been processed. But often there are model mapping relased publicly after the modesl were alread used (for example for testing or pre-sales). If you re-process the data with the new model mapping that represent the (i.e. meta-data) - you will get different aggregate, bot of them (old an new) are correct at the time of processing but they are also different if you re-run them at different time. So they are not immutable, but if your replace old aggregate by the new one, idempotency property is kept.

This is not a very "strict" definition, but when you consider real world use cases it is quite easy to map idempotency of Airlfow into what you should do when you re-run.

**Reply by potiuk** [reply]
If you find this description useful and find that the docs are unclear about it - I encourage you (as explained in the other discussion) to contribute back a PR improving it.

**Reply by aamster** [reply]
Thanks for your response. I'm not sure what idempotent here means then. If it creates a datafile foo.txt, then when it reruns, it overwrites foo.txt (with potential changes) What would **not** idempotent be then? That it no longer produces foo.txt but instead produces bar.txt?

I should also add that there are use cases where data was generated and overwriting that data would be extremely bad. Airflow does not seem to handle this use case and would overwrite it. Even updating it with an improvement might be bad.

**Reply by potiuk** [reply]
Each operator chooses to handle idempotency in the way that it's best for it. If your specific operator requires immutability - so be it. When you re-run it, it should check if the file is there and decide not to overwrite it. If the operator (or the user who configures it in their DAG decides that regenerating the file is better (and still idempotent) so be it. Airflow expects each operator to be idempotent, but how idempotency is achieved is left to the implementation of the operator.

This is far more flexible and "real-world" friendly than outright forbidding to modify produced file. Even more - Airflow currently does not (though it is changing with  dataset-driven scheduling) know, nor control the file that is output of each operator - it is left to the discretion of each operator's impleementation.

The responsibility to make sure your DAG steps are idempotent are put on the operator develope and DAG developer.

It's all in your hands. This is the philosophy of Airflow - more flexibility and control in the hands of their users, less limitations.

**Reply by aamster** [reply]
Sure, I guess I was asking if I was missing something. Since I was comparing with Luigi which makes the opinionated decision that if a file already exists, then the task used to produce that file will never be run again, thus out of the box enforcing what seems to be the straightforward definition of idempotency. Thanks for your help!
